File: go/algorithms/token_bucket.go
package algorithms

import (
	"sync"
	"time"
)

type bucket struct {
	tokens     float64
	lastRefill time.Time
}

type TokenBucket struct {
	mu           sync.RWMutex
	buckets      map[string]*bucket
	capacity     float64
	refillRate   time.Duration
	refillAmount float64
}

func NewTokenBucket(capacity int, refillRate time.Duration, refillAmount float64) *TokenBucket {
	return &TokenBucket{
		buckets:      make(map[string]*bucket),
		capacity:     float64(capacity),
		refillRate:   refillRate,
		refillAmount: refillAmount,
	}
}

func (tb *TokenBucket) Allow(key string) bool {
	tb.mu.Lock()
	defer tb.mu.Unlock()

	now := time.Now()
	b, exists := tb.buckets[key]
	if !exists {
		b = &bucket{
			tokens:     tb.capacity,
			lastRefill: now,
		}
		tb.buckets[key] = b
	}

	timePassed := now.Sub(b.lastRefill)
	refills := float64(timePassed) / float64(tb.refillRate)
	
	if refills >= 1 {
		b.tokens = min(tb.capacity, b.tokens+(refills*tb.refillAmount))
		b.lastRefill = now
	}

	if b.tokens >= 1 {
		b.tokens--
		return true
	}

	return false
}

func (tb *TokenBucket) Reset(key string) {
	tb.mu.Lock()
	defer tb.mu.Unlock()
	delete(tb.buckets, key)
}

func (tb *TokenBucket) GetStats(key string) RateLimiterStats {
	tb.mu.RLock()
	defer tb.mu.RUnlock()

	now := time.Now()
	b, exists := tb.buckets[key]
	if !exists {
		return RateLimiterStats{
			Allowed:   true,
			Limit:     int(tb.capacity),
			Remaining: int(tb.capacity),
			ResetAt:   now.Add(tb.refillRate),
		}
	}

	timePassed := now.Sub(b.lastRefill)
	refills := float64(timePassed) / float64(tb.refillRate)
	currentTokens := min(tb.capacity, b.tokens+(refills*tb.refillAmount))

	nextRefillTime := b.lastRefill.Add(time.Duration(float64(tb.refillRate) * (refills + 1)))

	return RateLimiterStats{
		Allowed:   currentTokens >= 1,
		Limit:     int(tb.capacity),
		Remaining: int(currentTokens),
		ResetAt:   nextRefillTime,
	}
}

func min(a, b float64) float64 {
	if a < b {
		return a
	}
	return b
}

File: typescript/src/algorithms/token-bucket.ts
import { RateLimiter, RateLimiterStats } from './rate-limiter.interface';

interface Bucket {
  tokens: number;
  lastRefill: number;
}

export class TokenBucket implements RateLimiter {
  private buckets: Map<string, Bucket> = new Map();
  private capacity: number;
  private refillRate: number;
  private refillAmount: number;

  constructor(capacity: number, refillRate: number, refillAmount: number = 1) {
    this.capacity = capacity;
    this.refillRate = refillRate;
    this.refillAmount = refillAmount;
  }

  async allow(key: string): Promise<boolean> {
    const now = Date.now();
    let bucket = this.buckets.get(key);

    if (!bucket) {
      bucket = {
        tokens: this.capacity,
        lastRefill: now
      };
      this.buckets.set(key, bucket);
    }

    const timePassed = now - bucket.lastRefill;
    const refills = Math.floor(timePassed / this.refillRate);
    
    if (refills > 0) {
      bucket.tokens = Math.min(
        this.capacity,
        bucket.tokens + (refills * this.refillAmount)
      );
      bucket.lastRefill = now;
    }

    if (bucket.tokens >= 1) {
      bucket.tokens--;
      return true;
    }

    return false;
  }

  async reset(key: string): Promise<void> {
    this.buckets.delete(key);
  }

  async getStats(key: string): Promise<RateLimiterStats> {
    const bucket = this.buckets.get(key);
    const now = Date.now();
    
    if (!bucket) {
      return {
        allowed: true,
        limit: this.capacity,
        remaining: this.capacity,
        resetAt: new Date(now + this.refillRate)
      };
    }

    const timePassed = now - bucket.lastRefill;
    const refills = Math.floor(timePassed / this.refillRate);
    const currentTokens = Math.min(
      this.capacity,
      bucket.tokens + (refills * this.refillAmount)
    );

    const nextRefillTime = bucket.lastRefill + 
      ((refills + 1) * this.refillRate);

    return {
      allowed: currentTokens >= 1,
      limit: this.capacity,
      remaining: Math.floor(currentTokens),
      resetAt: new Date(nextRefillTime)
    };
  }
}